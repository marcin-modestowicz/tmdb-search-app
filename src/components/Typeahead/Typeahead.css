.root {
  position: relative;
  font-family: sans-serif;
  font-size: 1em;
  line-height: 2em;
}

.suggestions {
  position: absolute;
  left: 0;
  top: 2.25em;
  display: block;
  margin: 0;
  padding: 0;
  box-shadow: 0 0 0.25em silver;
  border-radius: 0.25em;
  max-height: 20em;
  z-index: 1000;
  list-style: none;
  background-color: white;
  overflow: auto;
  width: 100%;
}

.suggestion {
  padding: 0 0.5em;
  color: silver;
  cursor: pointer;
  text-align: left;
}

.suggestion:first-child {
  border-radius: 0.25em 0.25em 0 0;
}

.suggestion:last-child {
  border-radius: 0 0 0.25em 0.25em;
}

.suggestion:hover {
  background-color: royalBlue;
  color: white;
}

.commonPart {
  font-weight: bold;
}
