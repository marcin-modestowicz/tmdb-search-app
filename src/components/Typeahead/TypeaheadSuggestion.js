// @flow

import React, { Component } from "react";
import { observable } from "mobx";
import { observer } from "mobx-react";
import type { FieldValue } from "../../types";
import styles from "./Typeahead.css";

// Static element with role button can handle interactions, but there's a bug in eslint
// plugin https://github.com/evcohen/eslint-plugin-jsx-a11y/issues/214
/* eslint-disable jsx-a11y/no-static-element-interactions */

type Props = {
  suggestion: FieldValue,
  searchFor: string,
  onClick: (suggestion: FieldValue) => void
};

@observer
class TypeaheadSuggestion extends Component<Props> {
  props: Props;

  handleClick = () => {
    this.props.onClick(this.props.suggestion);
  };

  renderSuggestion = () => {
    const { suggestion: { label }, searchFor } = this.props;
    const commonPartStartIndex = label
      .toLowerCase()
      .search(searchFor.toLowerCase());
    let firstPart = label;
    let lastPart = "";
    let commonPart = "";

    // In case server returns data that indirectly matches searched string
    // (e.g. abbreviations)
    if (commonPartStartIndex !== -1) {
      const commonPartEndIndex = commonPartStartIndex + searchFor.length;
      firstPart = label.slice(0, commonPartStartIndex);
      lastPart = label.slice(commonPartEndIndex);
      commonPart = label.slice(commonPartStartIndex, commonPartEndIndex);
    }

    return (
      <div>
        {firstPart}
        <span className={styles.commonPart}>{commonPart}</span>
        {lastPart}
      </div>
    );
  };

  render() {
    const { suggestion } = this.props;

    return (
      <li
        role="button"
        className={styles.suggestion}
        onClick={this.handleClick}
      >
        {this.renderSuggestion()}
      </li>
    );
  }
}

export default TypeaheadSuggestion;
