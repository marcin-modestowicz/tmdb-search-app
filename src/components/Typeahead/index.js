// @flow
export { default } from "./Typeahead";
