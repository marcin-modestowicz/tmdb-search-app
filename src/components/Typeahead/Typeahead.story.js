// @flow

import React from "react";
import { observable } from "mobx";
import { Observer } from "mobx-react";
import { storiesOf } from "@storybook/react";
import { withKnobs, text, boolean } from "@storybook/addon-knobs";
import Typeahead from "./Typeahead";

const store = observable({
  value: "",
  onChange(value) {
    store.value = value;
  },
  dataSource(input) {
    if (input) {
      if (window.fetch) {
        // free autocomplete api, only for testing purposes
        const url = `https://autocomplete.clearbit.com/v1/companies/suggest?query=${input}`;
        return fetch(url)
          .then(response => response.json())
          .then(companies =>
            companies.map(company => ({
              label: company.name,
              value: company.name
            }))
          );
      }

      return new Promise(resolve => {
        setTimeout(() => {
          resolve([
            {
              value: "Stranger Things",
              label: "Stranger Things"
            },
            {
              value: "Doctor Strange",
              label: "Doctor Strange"
            },
            {
              value: "Strangers with Candy",
              label: "Strangers with Candy"
            },
            {
              value: "Stranger than Fiction",
              label: "Stranger than Fiction"
            }
          ]);
        }, 100);
      });
    }

    return Promise.resolve([]);
  }
});

storiesOf("Typeahead", module)
  .addDecorator(getStory => <Observer>{getStory}</Observer>)
  .addDecorator(withKnobs)
  .add("simple", () => (
    <Typeahead
      value={store.value}
      dataSource={store.dataSource}
      onChange={store.onChange}
      placeholder={text("Placeholder", "Search")}
    />
  ));
