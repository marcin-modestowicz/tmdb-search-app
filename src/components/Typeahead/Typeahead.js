// @flow

import React, { Component } from "react";
import { reaction, observable, action, computed } from "mobx";
import type { IPromiseBasedObservable } from "mobx-utils";
import { fromPromise, PENDING, FULFILLED } from "mobx-utils";
import { observer } from "mobx-react";
import type { FieldValue } from "../../types";
import TypeaheadSuggestion from "./TypeaheadSuggestion";
import TextInput from "../TextInput";
import Spinner from "./spinner.svg";
import styles from "./Typeahead.css";

type Props = {
  value: any,
  onChange?: (value: string) => void,
  onSelect?: (value: string) => void,
  dataSource: (value: string) => Promise<FieldValue[]>,
  placeholder?: string
};

@observer
class Typeahead extends Component<Props> {
  static Suggestion = TypeaheadSuggestion;

  @observable value: string = "";
  @observable isInputFocused: boolean = false;
  @observable isListFocused: boolean = false;
  @observable suggestionsRequest: ?IPromiseBasedObservable<any> = null;
  props: Props;

  constructor(props: Props) {
    super(props);

    this.value = props.value;

    reaction(
      () => this.value,
      async value => {
        if (!this.isValueSet) {
          this.suggestionsRequest = fromPromise(this.props.dataSource(value));
        }
      },
      {
        name: "Get suggestions",
        delay: 500
      }
    );
  }

  componentWillReceiveProps(props: Props) {
    this.value = props.value;
  }

  @computed
  get isValueSet(): boolean {
    return this.value === this.props.value;
  }

  @computed
  get isActive(): boolean {
    return this.isInputFocused || this.isListFocused;
  }

  @computed
  get isLoading(): boolean {
    const request = this.suggestionsRequest;
    return request != null && request.state === PENDING;
  }

  @computed
  get suggestions(): FieldValue[] {
    const request = this.suggestionsRequest;
    return request != null && request.state === FULFILLED ? request.value : [];
  }

  @computed
  get hasSuggestions(): boolean {
    return this.suggestions.length > 0;
  }

  @action
  handleInputBlur = () => {
    this.isInputFocused = false;

    if (!this.isActive) {
      if (this.props.onChange) {
        this.props.onChange(this.value);
      }
      this.resetSuggestions();
    }
  };

  @action
  handleInputFocus = () => {
    this.isInputFocused = true;
  };

  @action
  handleListBlur = () => {
    this.isListFocused = false;
  };

  @action
  handleListFocus = () => {
    this.isListFocused = true;
  };

  @action
  handleChange = (value: string) => {
    this.value = value;
    this.resetSuggestions();
  };

  @action
  handleSelect = (suggestion: FieldValue) => {
    this.value = suggestion.label;
    if (this.props.onSelect) {
      this.props.onSelect(suggestion);
    }
    this.resetSuggestions();
    this.isListFocused = false;
  };

  @action
  resetSuggestions = () => {
    this.suggestionsRequest = null;
  };

  render() {
    const { placeholder } = this.props;

    return (
      <div className={styles.root}>
        <TextInput
          value={this.value}
          placeholder={placeholder}
          onChange={this.handleChange}
          onFocus={this.handleInputFocus}
          onBlur={this.handleInputBlur}
          autocomplete="off"
          autocorrect="off"
          autocapitalize="off"
          spellcheck={false}
          icon={this.isLoading ? <Spinner /> : null}
        />
        {this.hasSuggestions && (
          <ul
            className={styles.suggestions}
            onMouseEnter={this.handleListFocus}
            onMouseLeave={this.handleListBlur}
          >
            {this.suggestions.map(suggestion => (
              <Typeahead.Suggestion
                key={suggestion.id}
                suggestion={suggestion}
                searchFor={this.value}
                onClick={this.handleSelect}
              />
            ))}
          </ul>
        )}
      </div>
    );
  }
}

export default Typeahead;
