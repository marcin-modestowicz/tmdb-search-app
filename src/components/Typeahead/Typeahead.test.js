// @flow
import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import { fromPromise, PENDING, FULFILLED } from "mobx-utils";
import Typeahead from "./Typeahead";

describe("Typeahead component", () => {
  let onChange;
  let onSelect;
  let dataSource;
  let typeahead;
  let instance;
  let resetSuggestionsMock;

  beforeEach(() => {
    onChange = jest.fn();
    onSelect = jest.fn();
    dataSource = jest.fn(() =>
      Promise.resolve([
        { id: "1", value: "test1", label: "Test1" },
        { id: "2", value: "test2", label: "Test2" }
      ])
    );
    typeahead = shallow(
      <Typeahead
        value="one"
        onChange={onChange}
        onSelect={onSelect}
        dataSource={dataSource}
        placeholder="Search"
      />
    );
    instance = typeahead.instance();
    resetSuggestionsMock = jest.spyOn(instance, "resetSuggestions");
  });

  test("should render without suggestions", () => {
    expect(toJson(typeahead)).toMatchSnapshot();
  });

  test("should render with suggestions", async () => {
    instance.suggestionsRequest = fromPromise(dataSource());

    await instance.suggestionsRequest;

    expect(toJson(typeahead)).toMatchSnapshot();
  });

  describe("method handleInputBlur", () => {
    test("should set isInputFocused to false", () => {
      instance.isInputFocused = true;
      instance.handleInputBlur();
      expect(instance.isInputFocused).toBeFalsy();
    });

    test("if list is focused should do nothing", () => {
      instance.isListFocused = true;
      instance.handleInputBlur();
      expect(onChange).not.toHaveBeenCalled();
      expect(resetSuggestionsMock).not.toHaveBeenCalled();
    });

    describe("if list is not focused", () => {
      beforeEach(() => {
        instance.isListFocused = false;
        instance.handleInputBlur();
      });

      test("should call onChange prop", () => {
        expect(onChange).toHaveBeenCalledWith(instance.value);
      });

      test("should reset suggestions", () => {
        expect(resetSuggestionsMock).toHaveBeenCalled();
      });
    });
  });

  test("method handleInputFocus should set isInputFocused to true", () => {
    instance.isInputFocused = false;
    instance.handleInputFocus();
    expect(instance.isInputFocused).toBeTruthy();
  });

  test("method handleListBlur should set isListFocused to false", () => {
    instance.isListFocused = true;
    instance.handleListBlur();
    expect(instance.isListFocused).toBeFalsy();
  });

  test("method handleListFocus should set isListFocused to true", () => {
    instance.isListFocused = false;
    instance.handleListFocus();
    expect(instance.isListFocused).toBeTruthy();
  });

  describe("method handleChange", () => {
    beforeEach(() => {
      instance.handleChange("text");
    });

    test("should set value", () => {
      expect(instance.value).toBe("text");
    });

    test("should reset suggestions", () => {
      expect(resetSuggestionsMock).toHaveBeenCalled();
    });
  });

  describe("method handleSelect", () => {
    let suggestion;

    beforeEach(() => {
      suggestion = { label: "Test1", value: "test1" };
      instance.isListFocused = true;
      instance.handleSelect(suggestion);
    });

    test("should set value", () => {
      expect(instance.value).toBe("Test1");
    });

    test("should call onSelect with suggestion", () => {
      expect(onSelect).toHaveBeenCalledWith(suggestion);
    });

    test("should set isListFocused to false", () => {
      expect(instance.isListFocused).toBeFalsy();
    });

    test("should reset suggestions", () => {
      expect(resetSuggestionsMock).toHaveBeenCalled();
    });
  });
});
