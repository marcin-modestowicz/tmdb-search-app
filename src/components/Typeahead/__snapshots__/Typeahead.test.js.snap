// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Typeahead component should render with suggestions 1`] = `
<div
  className="root"
>
  <TextInput
    autoFocus={false}
    autocapitalize="off"
    autocomplete="off"
    autocorrect="off"
    icon={null}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="Search"
    spellcheck={false}
    value="one"
  />
  <ul
    className="suggestions"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <TypeaheadSuggestion
      key="1"
      onClick={[Function]}
      searchFor="one"
      suggestion={
        Object {
          "id": "1",
          "label": "Test1",
          "value": "test1",
        }
      }
    />
    <TypeaheadSuggestion
      key="2"
      onClick={[Function]}
      searchFor="one"
      suggestion={
        Object {
          "id": "2",
          "label": "Test2",
          "value": "test2",
        }
      }
    />
  </ul>
</div>
`;

exports[`Typeahead component should render without suggestions 1`] = `
<div
  className="root"
>
  <TextInput
    autoFocus={false}
    autocapitalize="off"
    autocomplete="off"
    autocorrect="off"
    icon={null}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="Search"
    spellcheck={false}
    value="one"
  />
</div>
`;
